File: export.js
const mysql = require('mysql2/promise');
const createCsvWriter = require('csv-writer').createArrayCsvWriter;

async function main() {
    const t1 = Date.now()
    const csvWriter = createCsvWriter({
        header: ['NAME', 'LANGUAGE'],
        path: 'node_output.csv'
    });
    await csvWriter.writeRecords([
        ['id', 'qty', 'price', 'total', 'column_0', 'column_1', 'column_2', 'column_3', 'column_4', 'column_5', 'column_6',' column_7', 'column_8', 'column_9', 'column_10']
    ])

    const connection = await mysql.createConnection({
        host: 'localhost',
        user: 'root',
        password: '123456',
        database: 'benchmark_test'
    });
    
    
    let after_id = 0;
    const limit = 10000;
    while (true) {
        const [rows] = await connection.execute(`SELECT * FROM products WHERE id > ${after_id} ORDER BY id ASC LIMIT ${limit}`);
        if (!rows.length) break;
        await csvWriter.writeRecords(rows.map(i => ([
            i.id,
            i.qty,
            i.price,
            i.qty * i.price,
            i.column_0,
            i.column_1,
            i.column_2,
            i.column_3,
            i.column_4,
            i.column_5,
            i.column_6,
            i.column_7,
            i.column_8,
            i.column_9,
            i.column_10,
        ])))

        after_id = rows[rows.length - 1].id
    }

    await connection.end();

    const t2 = Date.now();

    console.log(t2-t1);
}

main()

File: export.go
package main

import (
	"database/sql"
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
	"time"

	_ "github.com/go-sql-driver/mysql"
)

func main() {
	t1 := time.Now().UnixMilli()
	// Open a database connection
	const DB_USER = "root"
	const DB_PASSWORD = "123456"
	const DB_HOST = "localhost"
	const DB_PORT = 3306
	const DB_NAME = "benchmark_test"
	db, err := sql.Open("mysql", fmt.Sprintf("%v:%v@tcp(%v:%v)/%v", DB_USER, DB_PASSWORD, DB_HOST, DB_PORT, DB_NAME))
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	last_id := 0
	limit := 10000

	file, err := os.Create("go_outout.csv")
	if err != nil {
		log.Fatal("Error creating CSV file:", err)
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	writer.Write([]string{"id", "qty", "price", "total", "column_0", "column_1", "column_2", "column_3", "column_4", "column_5", "column_6", "column_7", "column_8", "column_9", "column_10"})

	for {
		rows, _ := db.Query(fmt.Sprintf("SELECT * FROM products WHERE id > %v ORDER BY id ASC LIMIT %v", last_id, limit))
		defer rows.Close()
		has_row := false

		for rows.Next() {
			has_row = true
			var id, qty, price int
			var column_0, column_1, column_2, column_3, column_4, column_5, column_6, column_7, column_8, column_9, column_10 string

			rows.Scan(&id, &qty, &price, &column_0, &column_1, &column_2, &column_3, &column_4, &column_5, &column_6, &column_7, &column_8, &column_9, &column_10)

			writer.Write([]string{strconv.Itoa(id), strconv.Itoa(qty), strconv.Itoa(price), strconv.Itoa(qty * price), column_0, column_1, column_2, column_3, column_4, column_5, column_6, column_7, column_8, column_9, column_10})

			last_id = id
		}

		if !has_row {
			break
		}
	}

	t2 := time.Now().UnixMilli()

	fmt.Println(t2 - t1)

}
